compilers/concrete-compiler/compiler/lib/Runtime/simulation.cpp

```diff
@@ -183,16 +205,147 @@ void sim_wop_pbs_crt(
+
+  uint64_t result = lhs + rhs;
+
+  if (is_signed) {
+    // We shift left to discard the padding bit and only consider the message
+    // for easier overflow checking
+    int64_t lhs_signed = (int64_t)lhs << 1;
+    int64_t rhs_signed = (int64_t)rhs << 1;
+    if (lhs_signed > 0 && rhs_signed > INT64_MAX - lhs_signed)
+      printf(msg_f, loc);
+    else if (lhs_signed < 0 && rhs_signed < INT64_MIN - lhs_signed)
+      printf(msg_f, loc);
+  } else if (lhs > UINT63_MAX - rhs || result > UINT63_MAX) {
+    printf(msg_f, loc);
+  }
+  return result;
+}
+
+uint64_t sim_mul_lwe_u64(uint64_t lhs, uint64_t rhs, char *loc,
+                         bool is_signed) {
+  const char msg_f[] =
+      "WARNING at %s: overflow happened during multiplication in simulation\n";
+
+  uint64_t result = lhs * rhs;
+
+  if (is_signed) {
+    // We shift left to discard the padding bit and only consider the message
+    // for easier overflow checking
+    int64_t lhs_signed = (int64_t)lhs << 1;
+    int64_t rhs_signed = (int64_t)rhs << 1;
+    if (lhs_signed != 0 && rhs_signed > INT64_MAX / lhs_signed)
+      printf(msg_f, loc);
+    else if (lhs_signed != 0 && rhs_signed < INT64_MIN / lhs_signed)
+      printf(msg_f, loc);
+  } else if (rhs != 0 && (lhs > UINT63_MAX / rhs || result > UINT63_MAX)) {
+    printf(msg_f, loc);
+  }
+  return result;
+}
+
+// a copy of memref_encode_expand_lut_for_bootstrap but which encodes overflow
+// and sign info into the LUT. Those information should later be discarder by
+// the LUT function
 void sim_encode_expand_lut_for_boostrap(
-    uint64_t *out_allocated, uint64_t *out_aligned, uint64_t out_offset,
-    uint64_t out_size, uint64_t out_stride, uint64_t *in_allocated,
-    uint64_t *in_aligned, uint64_t in_offset, uint64_t in_size,
-    uint64_t in_stride, uint32_t poly_size, uint32_t output_bits,
-    bool is_signed) {
-  return memref_encode_expand_lut_for_bootstrap(
-      out_allocated, out_aligned, out_offset, out_size, out_stride,
-      in_allocated, in_aligned, in_offset, in_size, in_stride, poly_size,
-      output_bits, is_signed);
+    uint64_t *output_lut_allocated, uint64_t *output_lut_aligned,
+    uint64_t output_lut_offset, uint64_t output_lut_size,
+    uint64_t output_lut_stride, uint64_t *input_lut_allocated,
+    uint64_t *input_lut_aligned, uint64_t input_lut_offset,
+    uint64_t input_lut_size, uint64_t input_lut_stride, uint32_t poly_size,
+    uint32_t out_MESSAGE_BITS, bool is_signed) {
+
+  assert(input_lut_stride == 1 && "Runtime: stride not equal to 1, check "
+                                  "memref_encode_expand_lut_bootstrap");
+
+  assert(output_lut_stride == 1 && "Runtime: stride not equal to 1, check "
+                                   "memref_encode_expand_lut_bootstrap");
+
+  size_t mega_case_size = output_lut_size / input_lut_size;
+
+  assert((mega_case_size % 2) == 0);
+
+  // compute overflow bit
+  bool *overflow_info = (bool *)malloc(sizeof(bool) * output_lut_size);
```

Comment:
Use vector instead (but should not be necessary if you factorize the code)

Reply:
Sometimes I forget that I'm not writing C code